File: chapra_7th/ch18/lagrange_interp.f90
subroutine lagrange_interp(N, x, y, xx, res)
  implicit none
  integer :: N 
  real(8) :: x(0:N), y(0:N)
  real(8) :: xx, res
  !
  real(8) :: ss, pp
  integer :: i, j

  ss = 0.d0
  do i = 0,N
    pp = y(i)
    do j = 0,N
      if( i /= j ) then
        pp = pp*( xx - x(j) ) / ( x(i) - x(j) )
      endif
    enddo
    ss = ss + pp
  enddo
  res = ss
  return
end subroutine 

File: chapra_7th/ch18/lagrange_interp.py
def lagrange_interp(x, y, xx):
    assert(len(x) == len(y))
    N = len(x) - 1  # length of array is (N + 1)
    ss = 0.0
    for i in range(0,N+1):
        pp = y[i]
        for j in range(0,N+1):
            if i != j:
                pp = pp*( xx - x[j] ) / ( x[i] - x[j] )
        ss = ss + pp
    return ss
